src/data_constructors.jl

```diff
@@ -1,3 +1,61 @@
+"""
+    modelcols(t::AbstractTerm, data)
+
+Create a numerical "model columns" representation of data based on an
+`AbstractTerm`.  `data` can either be a whole table (a property-accessible
+collection of iterable columns or iterable collection of property-accessible
+rows, as defined by [Tables.jl](https://github.com/JuliaData/Tables.jl) or a
+single row (in the form of a `NamedTuple` of scalar values).
+"""
+modelcols(t::ContinuousTerm, d::NamedTuple) = copy.(d[t.sym])
+modelcols(t::CategoricalTerm, d::NamedTuple) = t.contrasts[d[t.sym], :]
+function modelcols(t::MatrixTerm, d::Tables.ColumnTable)
+    mat = reduce(hcat, [modelcols(tt, d) for tt in t.terms])
+    return reshape(mat, size(mat, 1), :)
+end
+
+modelcols(t::MatrixTerm, d::NamedTuple) = reduce(vcat, [modelcols(tt, d) for tt in t.terms])
+# two options here: either special-case ColumnTable (named tuple of vectors)
+# vs. vanilla NamedTuple, or reshape and use normal broadcasting
+function modelcols(t::InteractionTerm, d::NamedTuple)
+    return kron_insideout(*, (modelcols(term, d) for term in t.terms)...)
+end
+
+function modelcols(t::InteractionTerm, d::Tables.ColumnTable)
+    return row_kron_insideout(*, (modelcols(term, d) for term in t.terms)...)
+end
+
+vectorize(x::Tuple) = collect(x)
+vectorize(x::AbstractVector) = x
+vectorize(x) = [x]
```

Comment:
Would this also be expected behavior?
```
julia> vectorize((; a=[1]))
1-element Vector{NamedTuple{(:a,), Tuple{Vector{Int64}}}}:
 (a = [1],)
```
The implicit conversion is nice here, but might also be risky.

Reply:
This is why I am putting tests. This is some copied code (since LICENSING permits) from StatsModels.jl regarding the formula and the creation of the `modelcols`